doc/specs/#1595 - Suggestions UI/Suggestions-UI.md

```diff
@@ -0,0 +1,734 @@
+* If the Suggestions UI is too close to the bottom of the screen, we'll need it to open
+  "upwards", with the search box at the _bottom_ and the list extending above it
+* prevent it from switching to command-line mode
+* display tooltips / `TeachingTip`s / some secondary flyout with a description
+  of the suggestion (if provided)
+
+#### Completion sources
+
+The Suggestions UI will support suggestions from a variety of different
+"sources". As an example, consider the following actions:
+
+```json
+        { "command": { "action":"suggestions", "source": "commandHistory" } },
+        { "command": { "action":"suggestions", "source": "directoryHistory" } },
+
+        { "command": { "action":"suggestions", "source": "tasks" } },
+        { "command": { "action":"suggestions", "source": "local" } },
+
+        { "command": { "action":"suggestions", "source": ["local", "tasks", "commandHistory"] } },
+
+        { "command": { "action":"suggestions", "source": "Microsoft.Terminal.Extensions.BufferComplete" } },
+```
+
+Each of these `suggestions` actions would open the Suggestions UI with a
+different set of actions.
+
+* `commandHistory`: Use commands from this session, as identified via shell
+  integration. This won't be able to return any suggestions if the user has not
+  configured their shell to support shell integration sequences yet.
+* `directoryHistory`: Populate the list with a series of `cd {path}` commands,
+  where the paths are populated via shell integration. Paths are in MRU order.
+* `tasks`: Populate the list with all `sendInput` actions in the user's settings
+  file. The command structure should remain unchanged. For example, if they have
+  `sendInput` actions nested under a "git" command, then the "git" entry will
+  remain in this tasks view with their `sendInput` actions nested inside it. For
+  more details, see the [Tasks] spec.
+* `local`: Populate the list with tasks that are located in the CWD, in a file
+  named `.wt.json`. For more details, see the [Tasks] spec.
+* `Microsoft.Terminal.Extensions.BufferComplete`: As an example, this
+  demonstrates how an action might be authored to reference a suggestion source
+  from an extension[[2](#footnote-2)].
+
+Each of these different sources will build a different set of `Command`s,
+primarily populated with `sendInput` actions. We'll load those `Command`s into
+the Suggestions UI control, and open it at the text cursor.
+
+To drill in on a single example - the `commandHistory` source. In that
+particular case, the TerminalPage will query the active TermControl for a list
+of its recent commands. If it knows these (via shell integration), then the
+TerminalPage will use that list of commands to build a list of `sendInput`
+actions. Those will then get fed to the suggestions UI.
+
+Not listed above is [shell-driven autocompletion]. These aren't something that
+the Terminal can invoke all on its own - these are something the shell would

```

Comment:
I especially don't love `sendInput` `Command`s for shell-driven autocompletion, for a couple reasons.

* They allow the shell to break the input barrier (James' example of "an application updates your clipboard and tells you to right-click something, oops, it pastes in `Ctrl-C Ctrl-C rm -rf / Enter`)
* Given that our main currency everywhere else in the terminal is "pretend the user typed something", it feels like the shell shouldn't _need_ to tell us what to type back into them. They've got more context than we do.

Reply:
(like: PowerShell says "To complete `Get-C`, type in `hildItem`" and Terminal then has to say "okay champ, `h` `i` `l` `d` `I`..." feels like the wrong level of abstraction)

Reply:
Ever since this comment, I can't help but think the protocol should just be "We send back a {sequence with an index}" and the shell/cli app can read that back to do what it wants with the chosen option. But this sure ain't the shell completion protocol spec 😝